### DISCUSSION TITLE
Library for UriTemplate

### DISCUSSION THREAD

**Discussion by kicito** [jolie/jolie#422] [Ideas (Features, Fixes, etc.)]
I am not sure if we need this PR #421 anymore as I now find another way to resolve the issue on my project with uri containing dash(-). By not including the query variable in the template. Jolie will automatically append the unused values to querystring anyway.

Thus with this setting

```
type GetDataTagRequest {
    a : string
    "page-after"? : string // page-after goes to query string

    Authorization?: string

}

osc.getDataTags << {
    template = "api/{a}/data-tags"
    method = "GET"
    outHeaders.("Authorization") = "Authorization"
}
```
sending  value
```
Value:
                                                ├─── Authorization:string = "Bearer token"
                                                ├─── page-after:string = "1"
                                                ├─── a:string = "a"
```
will send
```
GET /api/agents/a/data-tags?page-after=1 HTTP/1.1
Authorization: Bearer token
```

which is what I needed.

I need this trick no matter which library we are using. Since `damnhandy` doesn't support dash(-) on variable ( having {?page-after} throws `The variable name page-after contains invalid characters`), and Spring's `UriTemplate` supports only simple variable expansion not the whole [rfc6570](https://datatracker.ietf.org/doc/html/rfc6570)

**Comment by fmontesi** [comment]
Thanks for the write-up!
So it's a matter of deciding between the two libraries from other viewpoints. Spring's is well-maintained but bigger.

Are there features that damnhandy's implementation supports that Spring's library doesn't? What do you mean by "simple variable expansion"? Is that one of such features?

**Reply by kicito** [reply]
damnhandy's supports majority, if not all, expansions described in rfc6570 (AFAIK {var} and {?var}).

Spring's supports only {var} for utils.UriTemplates. If one requires more flexibility should use [UriComponentsBuilder](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/util/UriComponentsBuilder.html) instead. It doesn't follow rfc6570

**Reply by fmontesi** [reply]
I see. So... keep `damnhandy` for now? :)

**Reply by kicito** [reply]
Yup! closing the PR
